src/blog/fearless-structured-concurrency.vmd

```diff
@@ -0,0 +1,542 @@
+
+ * two or more threads concurrently accessing a location of memory
+ * one or more of them is a write
+ * one or more of them is unsynchronized
+
+
+Let's add a "progress" counter to our above snippet, and see a data race bug in action:
+
+```c
+#include <math.h>
+#include <stdio.h>
+#include <omp.h>
+
+int main() {
+  int exponent = 3;
+
+  int results[5];
+
+  int numIterationsCompleted = 0;
+  
+  // Launch some threads and run in parallel!
+  #pragma omp parallel for
+  for (int i = 0; i < 5; i++) {
+    results[i] = pow(i, exponent); // some expensive calculation
+
+    // Read, add, and store
+    numIterationsCompleted = numIterationsCompleted + 1;
+    printf("Completed %d iterations!\n", numIterationsCompleted);
+  }
+
+  for (int i = 0; i < 5; i++) {
+    printf("%d to the %d'th power is %d!\n", i, exponent, results[i]);
+  }
+  return 0;
+}
+```stdout
+Completed 1 iterations!
+Completed 2 iterations!
+Completed 3 iterations!
+Completed 4 iterations!
+Completed 4 iterations!
+0 to the 3'th power is 0!
+1 to the 3'th power is 1!
+2 to the 3'th power is 8!
+3 to the 3'th power is 27!
+4 to the 3'th power is 64!
+```
+
+Look at that! There are two `Completed 4 iterations!` printed out. [# If this doesn't happen for you, try running it a few more times. It's rather fickle.] Why is that?
+
+
+It's because the line `numIterationsCompleted = numIterationsCompleted + 1;` has three steps:
+
+ * Load `numIterationsCompleted` into register A.
+ * Add 1 to register A.
+ * Store register A into `numIterationsCompleted`.
+
+
+If two threads are running in parallel, we might see this happen:
+
+ * 4th thread loads 3 from `numIterationsCompleted` into register A.
+ * 5th thread loads 3 from `numIterationsCompleted` into register B.
+ * 4th thread adds 1 to register A, it now contains 4.
+ * 5th thread adds 1 to register B, it now contains 4.
+ * 4th thread stores register A into `numIterationsCompleted`, it now contains 4.
+ * 5th thread stores register B into `numIterationsCompleted`, it now contains 4.
+
+
+The problem is that they didn't see each others add operations; they're both adding 1 to the old value 3, to get 4.
+
+If instead the 4th thread's store happened before the 5th thread's load, we'd have gotten the correct answer 5.
+
+
+This is a *data race* bugs. They are _very_ difficult to detect, because they depend on the scheduling of the threads.
+
+Luckily, we've figured out how to avoid these problems, with some "concurrency ground rules":
+
+ 1. Multiple threads can read the same data, if nobody can modify it.
+ 1. If a thread can read data that another thread can modify, the data must be wrapped in a mutex. [# More on mutexes below.]
+ 1. If data is only visible to one thread, it can access that data freely.
+
+
+We usually need proper fear, respect, and discipline to stick to these rules. But when a language enforces these rules, we have *fearless concurrency*.
```

Comment:
Maybe add a note that there are exceptions to these rules, and that they don't always apply, like with say atomics.

Reply:
Added atomics to above paragraph